src/FibonacciHeap.java

```diff
@@ -101,11 +101,52 @@ private void consolidate() {
 	 * 
 	 */
 	public void decreaseKey(HeapNode x, int diff) {
-		// TODO: By Itay.
-		// Please look at the comment in detatchNode
-		return; // should be replaced by student code
+		x.key = x.key - diff;
+		if (x.parent == null|| x.parent.key<= x.key) { 
+			// The decrease is legal with the heap logic 
+			return;
+		}
+		else  {
+			cascadeCut(x);
+		}
+		return; 
+	}
+	/**
+	 * 
+	 * Handle the cutting process.
+	 *
+	 */
+	private void cascadeCut(HeapNode x) {
+		if (x.parent == null) 
+			//Terminate cascade when we reach root
+			return;
+		HeapNode parent = x.parent;
+		if (x.parent.child == x) {
+			if (x.next == x) {
+				// Only child
+				x.parent.child = null;
+			} else {
+				// Connect parent to sibling
+				x.parent.child = x.next;
+			}
+		}
+		numCuts++;
+		x.parent.rank--;
+		x.parent = null;
+		// Detach node from its siblings
+		removeFromLinkedList(x);
+		addToRootList(x);
+		x.mark = false; 
+		// The mark of a new tree root is always false
+		if (x.parent.mark == false&&x.parent.parent != null) {
```

Comment:
Missing spacing around AND(&&)
```suggestion
		if (x.parent.mark == false && x.parent.parent != null) {
```